File: 274/274_hindex_on.c
#include <stdio.h>
#include <stdlib.h>

int hIndex(int* citations, int citationsSize) {
    // Step 1: Count the number of each citation
    int* freq = (int*)calloc(citationsSize + 1, sizeof(int));
    
    // Step 2: Count the citations
    for (int i = 0; i < citationsSize; i++) {
        (citations[i] >= citationsSize) ? freq[citationsSize]++ : freq[citations[i]]++;
    }

    // Step 3: Sum citations from the end to calculate the H-index
    int total = 0;
    for (int i = citationsSize; i >= 0; i--) {
        total += freq[i];  // Total papers with i or more citations
        if (total >= i) {
            free(freq);  // Free the dynamically allocated memory
            return i;
        }
    }
    free(freq);  // Free the dynamically allocated memory
    return 0;
}

int main() {
    int citations[] = {3, 0, 6, 1, 5};
    int size = sizeof(citations) / sizeof(citations[0]);
    int h_index = hIndex(citations, size);
    printf("%d", h_index);
}

File: 274/274_hindex_on.py
from typing import List

class Solution:
    def hIndex(self, citations: List[int]) -> int:
        n = len(citations)
        freq = [0] * (n + 1)

        # Build the frequency array
        for c in citations:
            if c >= n:
                freq[n] += 1
            else:
                freq[c] += 1

        # Accumulate citations and calculate the H-index
        total = 0
        for i in range(n, -1, -1):
            total += freq[i]
            if total >= i:
                return i
        return 0

if __name__ == '__main__':
    s = Solution()
    print(s.hIndex([3, 0, 6, 1, 5]))